pkg/process/metadata/parser/nodejs/nodejs.go

```diff
@@ -0,0 +1,57 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package nodejs wraps functions to guess service name for node applications
+package nodejs
+
+import (
+	"encoding/json"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+type packageName struct {
+	// Name of the package in the json file
+	Name string `json:"name,omitempty"`
+}
+
+// FindNameFromNearestPackageJSON finds the package.json walking up from the absFilePath.
+// if a package.json is found, returns the value of the field name if declared
+func FindNameFromNearestPackageJSON(absFilePath string) (bool, string) {
+	current := filepath.Dir(absFilePath)
+	up := filepath.Dir(current)
+	for run := true; run; run = current != up {
+		ok, value := maybeExtractServiceName(filepath.Join(current, "package.json"))
+		if ok {
+			return ok && len(value) > 0, value
+		}
+		current = up
+		up = path.Dir(current)
+	}
+	ok, value := maybeExtractServiceName(filepath.Join(current, "package.json")) // this is for the root folder
+	return ok && len(value) > 0, value
+
+}
+
+// maybeExtractServiceName return true if a package.json has been found and eventually the value of its name field inside.
+func maybeExtractServiceName(filename string) (bool, string) {
+	// we can use a reader and use a json decoder also to limit the memory used and generally speaking to be safe.
+	// However, package.json files are supposed to be small and Unmarshall offers validation
+	bytes, err := os.ReadFile(filename)
```

Comment:
that can be problematic - reading the entire file into memory can lead to OOM or a significant memory impact on the process.
it is better to read the file line by line

Reply:
I'm not sure the new method is better

I ran a benchmark

```go
type packageName struct {
	// Name of the package in the json file
	Name string `json:"name,omitempty"`
}

func BenchmarkNew(b *testing.B) {
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		reader, err := os.Open("./package.json")
		require.NoError(b, err)
		pn := packageName{}
		err = json.NewDecoder(reader).Decode(&pn)
		require.NoError(b, err)
		reader.Close()
	}
}

func BenchmarkOrig(b *testing.B) {
	b.ReportAllocs()
	reader, err := os.Open("./package.json")
	require.NoError(b, err)
	defer reader.Close()
	for i := 0; i < b.N; i++ {
		bytes, err := os.ReadFile("/home/guy/Downloads/package.json")
		require.NoError(b, err)

		pn := packageName{}
		require.NoError(b, json.Unmarshal(bytes, &pn))
	}
}
```

[package.json](https://github.com/DataDog/datadog-agent/files/14300213/package.json)

Reply:
Naively I expected to have a stream-fashion parsing by providing a reader. I changed with more manual parsing minimizing the allocations as much as my knowledge could.
It looks way better (BenchmarkManual is the new method I just committed to use)

```
BenchmarkNew
BenchmarkNew-10       	       7	 143432000 ns/op	67107630 B/op	      32 allocs/op
BenchmarkManual
BenchmarkManual-10    	    8653	    133149 ns/op	   53120 B/op	    2964 allocs/op
BenchmarkOrig
BenchmarkOrig-10      	       8	 138620432 ns/op	26149591 B/op	      14 allocs/op

```

The nb of mallocs are still high but the memory stay low and the execution seems way faster

Reply:
I mark as resolved since personally satisfied. Please reopen the thread if there are still important issues to solve around this